packages/flutter/lib/src/scheduler/binding.dart

```diff
@@ -954,20 +954,45 @@ mixin SchedulerBinding on BindingBase {
 
   int _debugFrameNumber = 0;
   String? _debugBanner;
-  bool _ignoreNextEngineDrawFrame = false;
+
+  // Whether the current engine frame needs to be postponed till after the
+  // warm-up frame.
+  //
+  // Engine may begin a frame in the middle of the warm-up frame because the
+  // warm-up frame is scheduled by timers while the engine frame is scheduled
+  // by platform specific frame scheduler (e.g. `requestAnimationFrame` on the
+  // web). When this happens, we let the warm-up frame finish, and postpone the
+  // engine frame.
+  bool _rescheduleAfterWarmUpFrame = false;
 
   void _handleBeginFrame(Duration rawTimeStamp) {
     if (_warmUpFrame) {
-      assert(!_ignoreNextEngineDrawFrame);
-      _ignoreNextEngineDrawFrame = true;
+      // "begin frame" and "draw frame" must strictly alternate. Therefore
+      // _rescheduleAfterWarmUpFrame cannot possibly be true here as it is
+      // reset by _handleDrawFrame.
+      assert(!_rescheduleAfterWarmUpFrame);
+      _rescheduleAfterWarmUpFrame = true;
       return;
     }
     handleBeginFrame(rawTimeStamp);
   }
 
   void _handleDrawFrame() {
-    if (_ignoreNextEngineDrawFrame) {
-      _ignoreNextEngineDrawFrame = false;
+    if (_rescheduleAfterWarmUpFrame) {
+      _rescheduleAfterWarmUpFrame = false;
+      // Reschedule in a timer to allow the draw-frame phase of the warm-up
+      // frame to finish.
+      Timer.run(() {
```

Comment:
should this be a post-frame callback rather than a timer?

Reply:
I was wary of using any method that relies on the frame lifecycle, given that the issue we're dealing here is precisely a violation of said lifecycle :smiley: Having said that, I just tried, and a post-frame callback seems to work fine. LMK what your preference is, and whether it has some advantages that I could capture in a test.

Reply:
A post-frame callback would make the semantics a little clearer since we want to run this at the end of the currently executing frame (which is the warm-up frame).

Reply:
Done.